Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -8,52 +8,102 @@
-
-        if ((string?)manifest["mediaType"] != DockerManifestV2)
-        {
-            throw new NotImplementedException($"Do not understand the mediaType {manifest["mediaType"]}");
+        var client = GetClient();
+        var initialManifestResponse = await GetManifest(reference);
+        
+        return initialManifestResponse.Content.Headers.ContentType?.MediaType switch {
+            DockerManifestV2 => await TryReadSingleImage(await initialManifestResponse.Content.ReadFromJsonAsync<ManifestV2>()),
+            DockerManifestListV2 => await TryPickBestImageFromManifestList(await initialManifestResponse.Content.ReadFromJsonAsync<ManifestListV2>(), runtimeIdentifier),
+            var unknownMediaType => throw new NotImplementedException($"The manifest for {name}:{reference} from registry {BaseUri} was an unknown type: {unknownMediaType}. Please raise an issue at https://github.com/dotnet/sdk-container-builds/issues with this message.")
+        };
+
+        async Task<HttpResponseMessage> GetManifest(string reference) {
+            var client = GetClient();
+            var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{name}/manifests/{reference}"));
+            response.EnsureSuccessStatusCode();
+            return response;
         }
 
-        JsonNode? config = manifest["config"];
-        Debug.Assert(config is not null);
-        Debug.Assert(((string?)config["mediaType"]) == DockerContainerV1);
+        async Task<HttpResponseMessage> GetBlob(string digest) {
+            var client = GetClient();
+            var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{name}/blobs/{digest}"));
+            response.EnsureSuccessStatusCode();
+            return response;
+        }
 
-        string? configSha = (string?)config["digest"];
-        Debug.Assert(configSha is not null);
+        async Task<Image?> TryReadSingleImage(ManifestV2 manifest) {
+            var config = manifest.config;
+            string configSha = config.digest;
+            
+            var blobResponse = await GetBlob(configSha);
 
-        response = await client.GetAsync(new Uri(BaseUri, $"/v2/{name}/blobs/{configSha}"));
+            JsonNode? configDoc = JsonNode.Parse(await blobResponse.Content.ReadAsStringAsync());
+            Debug.Assert(configDoc is not null);
 
-        JsonNode? configDoc = JsonNode.Parse(await response.Content.ReadAsStringAsync());
-        Debug.Assert(configDoc is not null);
-        //Debug.Assert(((string?)configDoc["mediaType"]) == DockerContainerV1);
+            return new Image(manifest, configDoc, name, this);
+        }
 
-        return new Image(manifest, configDoc, name, this);
+        async Task<Image?> TryPickBestImageFromManifestList(ManifestListV2 manifestList, string runtimeIdentifier) {
+            // TODO: we probably need to pull in actual RID parsing code and look for 'platform' here.
```

Comment:
To do this correctly we should use the `RuntimeGraph` from Nuget.Packaging. We'll need the actual rid graph json to serve as the source of truth about the relationships, and we may also need to construct a sub-RID-graph from the manifestList manifests, then reconcile the two.